Check for expired transactions and batches

PhilippGackstatter opened the pull request:
## Changes

- Batches: Check that none of the transactions in a batch are expired and tests expired and unexpired transactions.
- Batches: Removes the `BatchNoteTree` from `ProposedBatch` and `ProvenBatch`.
- Block: Check that the batches included in the block are unexpired and tests expired and unexpired batches.
- Rename `ProvenBatch::new` -> `ProvenBatch::new_unchecked` to match `ProvenBlock::new_unchecked` (both of which don't do any validation).
- Simplify the transaction reference block check in the batch to avoid allocating a new `BTreeSet`.
- Rename `BatchProveError` to `ProvenBatchError` to align with `ProvenBlockError`, `ProposedBlockError` and `ProposedBatchError`.

closes #1156
closes #1159
closes #920

bobbinth commented on crates/miden-objects/src/batch/proven_batch.rs:
Should we also remove `output_notes_smt` from `ProvenBatch` in this PR? Or leave it for another PR?

PhilippGackstatter replied:
Ah yes, let's do that!

bobbinth reviewed (approved):
Looks good! Thank you! I primarily reviewed non-test code in this review.

Mirko-von-Leipzig commented on crates/miden-block-prover/src/tests/proposed_block_success.rs:
```diff
@@ -194,3 +195,32 @@ fn proposed_block_authenticating_unauthenticated_notes() -> anyhow::Result<()> {
+
+/// Tests that a batch that expires at the block being proposed is still accepted.
+#[test]
+fn proposed_block_unexpired_batches() -> anyhow::Result<()> {
```
Maybe 🤷‍♂️ 
```suggestion
fn proposed_block_with_batch_at_expiration_limit() -> anyhow::Result<()> {
```

Mirko-von-Leipzig commented on crates/miden-block-prover/src/tests/utils.rs:
```diff
@@ -140,34 +140,58 @@ pub fn generate_tx_with_authenticated_notes(
 }
 
+/// Generates a transaction that expires at the given block number.
+pub fn generate_tx_with_expiration(
+    chain: &mut MockChain,
+    account_id: AccountId,
+    expiration_block: BlockNumber,
+) -> ProvenTransaction {
```
Unrelated: what's the reason all these are standalone functions?

PhilippGackstatter replied:
I was in a rush to finish the tests, so I didn't put too much effort into the test utils. At some point in the middle of it I thought it would be nicer if this was an extension trait but didn't want to spend time on refactoring. Another option is to integrate them into the MockChain directly (but that has to be looked at in isolation and weighed against the overall API of the MockChain) and they're probably too purpose-built for that anyway.

Mirko-von-Leipzig replied:
I was thinking of the latter, but more in the builder style, e.g.

```rust
chain
  .open_block(..)
    .add_tx()
      .with_account(..)
      .with_expiration(..)
      .finalise()
    .seal();
```

PhilippGackstatter replied:
Nice idea. I guess this would have "sub-builders" for transactions and batches so that we can craft individual transactions, batches or blocks for whatever is needed for a given test.

Mirko-von-Leipzig commented on crates/miden-objects/src/batch/proposed_batch.rs:
```diff
@@ -146,19 +146,14 @@ impl ProposedBatch {
         }
 
-        // Verify all block references from the transactions are in the chain.
+        // Verify all block references from the transactions are in the chain MMR, except for the
+        // batch's reference block.
```
Could you explain the why for this

PhilippGackstatter replied:
Here or in code :D?

A batch references a previous block against which note inclusion is proven or that the reference blocks of transactions were included. In case that previous block of a batch is the latest block in the chain, it won't be included in the blockchain MMR yet. The reason for this is the same as we've discussed yesterday, namely that a block only updates the chain MMR with the previous block, but not itself. So in that sense, the latest block is a special case and its inclusion cannot be proven in the chain MMR. Instead, the MASM batch kernel also will commit to its reference block as a public input and assume it is valid. The block will then verify that all reference blocks of batches are in the chain (or if they are the latest block at the time of block building, then that it matches that block's commitment/hash).

For transactions, this is basically the same. The transaction kernel adds the block it references to the MMR within the prologue of the transaction, so that note inclusion proofs can be verified without checking for this special case:

https://github.com/0xPolygonMiden/miden-base/blob/dca16fd814bc7f7ef53c57a6edf520720e8fdfdd/crates/miden-lib/asm/kernels/transaction/lib/prologue.masm#L297-L310

And the transaction also has the hash of the block it references as a public input, so that the batch can verify its inclusion. Due to that verification chain up to the block, we eventually know that the reference blocks of transactions or batches were indeed included.
Hope that made some sense.

Mirko-von-Leipzig replied:
I think it definitely deserves a comment in code because naively one would assume that we should be proving the chain MMR up to and including the highest block referred to by the batch. And current comment says literally that one is excluded on purpose but not why 🙃 

----------------------

1. What is the batch's reference block? Is it the maximum of all its contents' reference blocks? Is it the block height at the time of batch "conception"? Or more likely it doesn't matter?
2. Given that the latest block cannot be proven, does this mean content reference blocks are also skipped if they don't have an inclusion proof? I think you've said as much.

PhilippGackstatter replied:
I'll address your comment in #1172 so this can be merged.

PhilippGackstatter commented:
I pushed a commit and only then saw that you already approve the PR, so I force pushed the previous state and added the commit addressing the comments. I'll make a separate PR for the empty block changes.
